File: tests/human/x/ex/group_by_conditional_sum.ex
items = [
  %{cat: "a", val: 10, flag: true},
  %{cat: "a", val: 5, flag: false},
  %{cat: "b", val: 20, flag: true}
]

result =
  items
  |> Enum.group_by(& &1.cat)
  |> Enum.sort_by(fn {k, _} -> k end)
  |> Enum.map(fn {cat, group} ->
    total = Enum.sum(Enum.map(group, & &1.val))
    share =
      Enum.map(group, fn x -> if x.flag, do: x.val, else: 0 end)
      |> Enum.sum()
      |> Kernel./(total)
    %{cat: cat, share: share}
  end)

IO.inspect(result)

File: tests/machine/x/fs/group_by_conditional_sum.fs
open System

exception Break
exception Continue

type Anon1 = {
    cat: string
    ``val``: int
    flag: bool
}
type Anon2 = {
    cat: obj
    share: obj
}
let items = [{ cat = "a"; ``val`` = 10; flag = true }; { cat = "a"; ``val`` = 5; flag = false }; { cat = "b"; ``val`` = 20; flag = true }]
let result = [ for gKey, gItems in [ for i in items do yield i ] |> List.groupBy (fun i -> i.cat) |> List.sortBy (fun (gKey, gItems) -> let g = {| key = gKey; items = gItems |} in g.key) do
    let g = {| key = gKey; items = gItems |}
    yield { cat = g.key; share = List.sum [ for x in g do yield (if x.flag then x.val else 0) ] / List.sum [ for x in g do yield x.val ] } ]
printfn "%A" (result)
